src/insights_client/__init__.py

```diff
@@ -208,47 +230,62 @@ def run_phase(phase, client, validated_eggs):
-    ] + sys.argv[1:]
     config = client.get_conf()
 
-    all_eggs = [ENV_EGG, NEW_EGG] + validated_eggs
+    all_eggs = [NEW_EGG] + validated_eggs
+    if ENV_EGG is not None:
+        all_eggs = [ENV_EGG] + all_eggs
+        logger.debug("Environment egg defined as %s." % ENV_EGG)
+
+    for egg in all_eggs:
+        if config["gpg"]:
+            if not os.path.isfile(egg):
+                logger.debug("%s is not a file, can't GPG verify. Skipping.", egg)
+                continue
+            if not client.verify(egg)["gpg"]:
+                logger.debug("Not using egg: %s", egg)
+                continue
+        else:
+            logger.debug("GPG disabled by --no-gpg, not verifying %s.", egg)
 
-    for i, egg in enumerate(all_eggs):
-        if egg is None or (config["gpg"] and not os.path.isfile(egg)):
-            client_debug("Egg does not exist: %s" % egg)
-            continue
-        if config["gpg"] and not client.verify(egg)["gpg"]:
-            client_debug("WARNING: GPG verification failed. Not loading egg: %s" % egg)
-            continue
-        client_debug("phase '%s'; egg '%s'" % (phase["name"], egg))
+        logger.debug("phase '%s'; egg '%s'", phase["name"], egg)
 
         # prepare the environment
+        env = os.environ
+        logger.debug("Existing enviroment: %s", debug_environ(env))
+
         pythonpath = str(egg)
-        env_pythonpath = os.environ.get("PYTHONPATH", "")  # type: str
+        logger.debug("Adding to PYTHONPATH: %s", pythonpath)
+        env_pythonpath = env.get("PYTHONPATH", "")  # type: str
+
         if env_pythonpath:
             pythonpath += ":" + env_pythonpath
         insights_env = {
             "INSIGHTS_PHASE": str(phase["name"]),
             "PYTHONPATH": pythonpath,
         }
-        env = os.environ
+        logger.debug("Adding to the environment: %s", debug_environ(insights_env))
+
         env.update(insights_env)
 
+        current_path = os.path.dirname(__file__)
+        logger.debug("Current path: %s", current_path)
+
+        insights_command = [
+            sys.executable,
+            os.path.join(current_path, "run.py"),
+        ] + sys.argv[1:]
+        logger.debug("Running command: %s", debug_command(insights_command, env))
         process = subprocess.Popen(insights_command, env=env)
-        stdout, stderr = process.communicate()
+        process.communicate()
         if process.returncode == 0:
             # phase successful, don't try another egg
-            client_debug("phase '{phase}' successful".format(phase=phase["name"]))
+            logger.debug("phase '%s' successful", phase["name"])
             update_motd_message()
             return
 
-        client_debug(
-            "phase '{phase}' failed with return code {rc}".format(
-                phase=phase["name"], rc=process.returncode
-            )
+        logger.debug(
+            "phase '%s' failed with return code %d", phase["name"], process.returncode
         )
```

Comment:
Actually, looking at Core, return code 100 means "OK, do not continue". There should be a condition that separates 0 and 100 from other exit codes.
```py
if process.returncode not in (0, 100):
    logger.debug(...)
```

Reply:
Don’t you consider the actual return code a useful piece of information? With a condition like this we don’t know whether the phase exitted with 0 or 100. Actually, we can know that if we inspect the source code, because in case of 0, the “phase '%s' successful” is logged but for 100 it’s not. 🤷🏻‍♂️

Reply:
Added the condition you suggested.

— A little side note: I used a list instead of a tuple. Although I know that tuples are more efficient and convey a meaning of an immutable object, I’d rather use them semantically as a real _tuple_ – a fixed set of coupled values, a structure if you will. In this case, we _list_ out the exit code for the condition.